### ISSUE TITLE
Update for PHP 8 Compatibility

### ISSUE THREAD

**Issue by getsource** [WordPress/phpunit-test-runner#121]
We should test the test runner for PHP 8 compatibility so that it's ready when core is compatible.

**Comment by getsource** [comment]
It looks like this is going to require optionally using a different version of PHPUnit.

I haven't got it to work successfully yet, but the [code in question is here](https://github.com/WordPress/phpunit-test-runner/blob/master/prepare.php#L35).

**Comment by getsource** [comment]
A quick note that `prepare.php` can be run on a different server and may have different PHP available from where the tests are actually run, so we should be sure that the correct version of PHPUnit gets installed for the destination PHP version.

**Comment by javiercasares** [comment]
inside functions.php

```
funtion detect_phpunit_file() {
  if(version_compare(PHP_VERSION, '8.0') >= 0 ) {
    return 'https://phar.phpunit.de/phpunit-9.phar';
  } else {
    return 'https://phar.phpunit.de/phpunit-5.7.phar';
  }
  return 'https://phar.phpunit.de/phpunit-5.7.phar';
}
```
update the line on prepare

`'wget -O ' .  escapeshellarg( $WPT_PREPARE_DIR . '/phpunit.phar' ) . ' ' . detect_phpunit_file(),`

about the PHP_VERSION, we should try the get_env_details()` that returns `$env = array( 'php_version'    => phpversion(),` so change it with` $env['php_version']` if we can access that.

**Comment by getsource** [comment]
I've been sick and not able to get to this, but wanted to be sure to pass on this info, since it should help:

I asked if it's possible to use Composer to install based on a different version of PHP that is not the current one on the system.

From @jrfnl: https://wordpress.slack.com/archives/C02RQBWTW/p1605101934303000?thread_ts=1605072913.278300&cid=C02RQBWTW
> Yes, you can use the platform key for that, like so:
```
 "config": {
    "platform": {
      "php": "5.6.40"
    }
  },
```
@jftnl also noted this, which I think is the best option:
> You can also set platform from the command-line (or overrule it) if needs be via a script. 

This would require composer on the prepare side (I don't believe it is currently required), but that should be fine as long as test reporters are notified ahead of time.

Putting it all together, we can:
In `prepare` step:
- Retrieve the version of PHP from the test environment, whether locally/SSH, depending on config
- Set composer `platform` to that version of PHP
- Install PHPUnit with composer
- PHPUnit gets rsync'd along with the rest of the prepared environment

In `test` step:
- Update `test.php` to use the version of PHPUnit that gets installed with composer

**Comment by jrfnl** [comment]
Additional note based on the above discussion:

The WP Core test suite is **NOT** compatible with PHPUnit 8/9 and will actively block itself from being run with any PHPUnit version > 7.5. A hack has been used to get round the PHPUnit 7 incompatibility with PHP 8 (WP local copy of some of the PHPUnit 9 files which overload the same files from PHPUnit 7), so the test suite will run on PHP 8 using PHPUnit 7.5.20, but ONLY if installed via Composer.
The Phar will not work as PHPUnit 7 is no longer supported and the last Phar generated contains dependencies of PHPUnit which are incompatible with PHP 8. If PHPUnit is installed via Composer, the most recent version of those dependencies compatible with PHPUnit 7 will be installed, which will allow the tests to run on PHP 8 with PHPUnit 7.

**Comment by ottok** [comment]
Forcing phpunit 7.5 to install despite its own `^7.1` restriction was the solution. Works now:
![image](https://user-images.githubusercontent.com/668724/100478713-c0d0e080-30f4-11eb-8995-be91b28cf6ee.png)

**Comment by getsource** [comment]
I chatted with @jrfnl (thank you!!) and we came up with the following steps to make this compatible with PHP 5.6->8.X:
- Detect PHP 5.6-7.0, and use PHPUnit 5 for it, like is currently happening in master, because these versions are not currently supported by Core's composer configuration.
- If PHP 8 is detected, instead force PHP 7.4 as the version. This is because:
    - PHPUnit will refuse to install for PHP 8, but works when the necessary dependencies are installed via composer. 
    - Core tests currently are set not to run on newer versions of PHPUnit that are compatible with PHP 8.
- Finally, use composer to install the most appropriate PHPUnit version.

**Comment by getsource** [comment]
Fixed by: #132

Here are the currently running and not-currently-reporting hosts, so that we can pay attention if anything breaks:
```
Active Test Reporters

    A2 Hosting
    Altis by Human Made
    AMIMOTO
    Conetix
    CynderHost
    DreamHost
    InMotion Hosting Test Bot
    Kinsta
    ManagedWPHosting
    Morgan Digital
    redIT
    redIT
    SAKURA internet
    Seravo
    SHIFTER
    SiteGround Bot
    Tuonetti
    Wetopi
    WordPress.org GitHub Actions
    XetNET

Registered, but no reports in >25 Revisions

    Bluehost
    Cloudways
    GoDaddy
    ICDSoft
    IONOS
    Nexcess
    Onyx
    Pantheon
    RAIDBOXES
    Savvii
    Touchstone Bot
    WP-Cloud
    WPMU DEV
    WPWebHost
    XSERVER
    XSERVER
    ZNETLIVE
```
